Yesod/Auth/OAuth2.hs

```diff
@@ -52,28 +54,46 @@ authOAuth2 name oauth getCreds = AuthPlugin name dispatch login
     where
         url = PluginR name ["callback"]
 
-        withCallback = do
+        withCallback csrfToken = do
             tm <- getRouteToParent
             render <- lift $ getUrlRender
-            return $ oauth { oauthCallback = Just $ encodeUtf8 $ render $ tm url }
+            let newEndpoint = oauthOAuthorizeEndpoint oauth `BS.append` "&state=" `BS.append` encodeUtf8 csrfToken
```

Comment:
The state parameter is recommended according to section 4.1.1 in the [RFC](http://tools.ietf.org/html/rfc6749). If you're worried about sites that don't support it, we could also add a `authOAuth2NoState` function for compatibility?